pkg/controller/namespace/namespace_controller.go

```diff
@@ -209,25 +209,58 @@ func deleteAllContent(kubeClient client.Interface, experimentalMode bool, namesp
 	return estimate, nil
 }
 
-// syncNamespace makes namespace life-cycle decisions
-func syncNamespace(kubeClient client.Interface, experimentalMode bool, namespace api.Namespace) (err error) {
-	if namespace.DeletionTimestamp == nil {
-		return nil
+// updateNamespaceFunc is a function that makes an update to a namespace
+type updateNamespaceFunc func(kubeClient client.Interface, namespace *api.Namespace) (*api.Namespace, error)
+
+// retryOnConflictError retries the specified fn if there was a conflict error
+// TODO RetryOnConflict should be a generic concept in client code
+func retryOnConflictError(kubeClient client.Interface, namespace *api.Namespace, fn updateNamespaceFunc) (result *api.Namespace, err error) {
+	result = namespace
+	latestNamespace := namespace
+	for true {
+		result, err = fn(kubeClient, latestNamespace)
+		if err == nil {
+			break
+		}
+		if !errors.IsConflict(err) {
+			return nil, err
+		}
+		latestNamespace, err = kubeClient.Namespaces().Get(latestNamespace.Name)
+		if err != nil {
+			return nil, err
+		}
 	}
-	glog.V(4).Infof("Syncing namespace %s", namespace.Name)
+	return result, nil
+}
 
-	// if there is a deletion timestamp, and the status is not terminating, then update status
+// updateNamespaceStatusFunc will verify that the status of the namespace is correct
+func updateNamespaceStatusFunc(kubeClient client.Interface, namespace *api.Namespace) (*api.Namespace, error) {
+	// only update the status to terminating if we have a deletion timestamp
 	if !namespace.DeletionTimestamp.IsZero() && namespace.Status.Phase != api.NamespaceTerminating {
 		newNamespace := api.Namespace{}
 		newNamespace.ObjectMeta = namespace.ObjectMeta
 		newNamespace.Status = namespace.Status
 		newNamespace.Status.Phase = api.NamespaceTerminating
-		result, err := kubeClient.Namespaces().Status(&newNamespace)
-		if err != nil {
-			return err
+		return kubeClient.Namespaces().Status(&newNamespace)
+	}
+	return namespace, nil
+}
+
+// syncNamespace orchestrates deletion of a Namespace and its associated content.
+func syncNamespace(kubeClient client.Interface, experimentalMode bool, namespace *api.Namespace) (err error) {
+	if namespace.DeletionTimestamp == nil {
+		return nil
+	}
+	glog.V(4).Infof("Syncing namespace %s", namespace.Name)
+
+	// ensure that the status is up to date on the namespace
+	// if we get a not found error, we assume the namespace is truly gone
+	namespace, err = retryOnConflictError(kubeClient, namespace, updateNamespaceStatusFunc)
```

Comment:
Yeah, I thought of it like append so it will return a new pointer rather than modify the incoming object.  I didn't want to modify the incoming object because its ultimately derived from the backing cache.DeltaFifo which is supposed to have read-only objects.